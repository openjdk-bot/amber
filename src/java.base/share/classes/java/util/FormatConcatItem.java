/*
 * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

package java.util;

import java.util.FormatItem.FormatItemDecimal;
import java.util.FormatItem.FormatItemHexadecimal;
import java.util.FormatItem.FormatItemOctal;
import java.util.FormatItem.FormatItemBoolean;
import java.util.FormatItem.FormatItemCharacter;
import java.util.FormatItem.FormatItemString;
import java.util.FormatItem.FormatItemFormatSpecifier;
import java.util.FormatItem.FormatItemModifier;
import java.util.FormatItem.FormatItemFillLeft;
import java.util.FormatItem.FormatItemFillRight;
import java.util.FormatItem.FormatItemUpper;
import java.util.FormatItem.FormatItemNull;

import jdk.internal.javac.PreviewFeature;

/**
 * Implementations of this class provide information necessary to
 * assist {@link java.lang.invoke.StringConcatFactory} perform optimal
 * insertion.
 *
 * @since 20
 */
@PreviewFeature(feature=PreviewFeature.Feature.STRING_TEMPLATES)
public sealed interface FormatConcatItem
    permits FormatItemDecimal,
            FormatItemHexadecimal,
            FormatItemOctal,
            FormatItemBoolean,
            FormatItemCharacter,
            FormatItemString,
            FormatItemFormatSpecifier,
            FormatItemModifier,
            FormatItemFillLeft,
            FormatItemFillRight,
            FormatItemUpper,
            FormatItemNull
{
    /**
     * Calculate the length of the insertion.
     *
     * @param lengthCoder current value of the length + coder
     * @return adjusted value of the length + coder
     */
    long mix(long lengthCoder);

    /**
     * Insert content into buffer prior to the current length.
     *
     * @param lengthCoder current value of the length + coder
     * @param buffer      buffer to right into
     *
     * @return adjusted value of the length + coder
     *
     * @throws Throwable if fails to prepend value (unusual).
     */
    long prepend(long lengthCoder, byte[] buffer) throws Throwable;
}
