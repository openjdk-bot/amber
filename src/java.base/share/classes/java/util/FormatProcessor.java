/*
 * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

package java.util;

import java.lang.invoke.*;
import java.lang.invoke.MethodHandles.Lookup;
import java.lang.template.*;
import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import jdk.internal.javac.PreviewFeature;

/**
 * This {@link TemplateProcessor} constructs a String result using {@link
 * Formatter}. Unlike {@link Formatter}, FormatProcessor uses the value from
 * the embedded expression that follows immediately after the
 * <a href="../util/Formatter.html#syntax">format specifier</a>.
 * StringTemplate expressions without a preceeding specifier, use "%s" by
 * default. Example:
 * {@snippet :
 * int x = 10;
 * int y = 20;
 * String result = FMT."%05d\{x} + %05d\{y} = %05d\{x + y}";
 * }
 * result is: <code>00010 + 00020 = 00030</code>
 *
 * @implNote When used in conjunction with a compiler generated {@link
 * StringTemplate} this {@link TemplateProcessor} will use the format
 * specifiers in the fragments and types of the values in the value list
 * to produce a more performant formatter.
 *
 * @implSpec Since, values are in situ, argument indexing is unsupported.
 *
 * @since 20
 */
@PreviewFeature(feature=PreviewFeature.Feature.STRING_TEMPLATES)
public final class FormatProcessor implements StringProcessor, ProcessorLinkage {
    /**
     * {@link Locale} used to format
     */
    private final Locale locale;

    /**
     * Constructor.
     *
     * @param locale  {@link Locale} used to format
     */
    public FormatProcessor(Locale locale) {
        this.locale = locale;
    }

    /**
     * {@inheritDoc}
     * @throws  IllegalFormatException
     *          If a format string contains an illegal syntax, a format
     *          specifier that is incompatible with the given arguments,
     *          insufficient arguments given the format string, or other
     *          illegal conditions.  For specification of all possible
     *          formatting errors.
     * @see java.util.Formatter
     */
    @Override
    public final String process(StringTemplate stringTemplate) {
        Objects.requireNonNull(stringTemplate);
        String format = stringTemplateFormat(stringTemplate.fragments());
        Object[] values = stringTemplate.values().toArray(new Object[0]);

        return new Formatter(locale).format(format, values).toString();
    }

    /**
     * {@inheritDoc}
     * @throws  IllegalFormatException
     *          If a format string contains an illegal syntax, a format
     *          specifier that is incompatible with the given arguments,
     *          insufficient arguments given the format string, or other
     *          illegal conditions.  For specification of all possible
     *          formatting errors.
     * @see java.util.Formatter
     */
    @Override
    public MethodHandle linkage(List<String> fragments, MethodType type) {
        Objects.requireNonNull(fragments);
        Objects.requireNonNull(type);
        String format = stringTemplateFormat(fragments);
        Class<?>[] ptypes = type.dropParameterTypes(0, 1).parameterArray();
        MethodHandle mh = new FormatterBuilder(format, locale, ptypes).build();
        mh = MethodHandles.dropArguments(mh, 0, type.parameterType(0));

        return mh;
    }

    // %[argument_index$][flags][width][.precision][t]conversion
    private static final String formatSpecifier
            = "%(\\d+\\$)?([-#+ 0,(\\<]*)?(\\d+)?(\\.\\d+)?([tT])?([a-zA-Z%])";

    private static final Pattern fsPattern = Pattern.compile(formatSpecifier);

    /**
     * Find a format specification at the end of a fragment.
     *
     * @param fragment  fragment to check
     * @param needed    if the specification is needed
     *
     * @return true if the specification is found and needed
     *
     * @throws MissingFormatArgumentException if not at end or found and not needed
     */
    private static boolean findFormat(String fragment, boolean needed) {
        Matcher matcher = fsPattern.matcher(fragment);
        String group;

        while (matcher.find()) {
            group = matcher.group();

            if (!group.equals("%%") && !group.equals("%n")) {
                if (matcher.end() == fragment.length() && needed) {
                    return true;
                }

                throw new MissingFormatArgumentException(group);
            }
        }

        return false;
    }

    /**
     * Convert a {@link StringTemplate} fragments, containing format specifications,
     * to a form that can be passed on to {@link Formatter}. The method scans each fragment,
     * matching up formatter specifications with the following expression. If no
     * specification is found, the method inserts "%s".
     *
     * @param fragments  string template fragments
     *
     * @return  format string
     */
    private static String stringTemplateFormat(List<String> fragments) {
        StringBuilder sb = new StringBuilder();
        int lastIndex = fragments.size() - 1;
        List<String> formats = fragments.subList(0, lastIndex);
        String last = fragments.get(lastIndex);

        for (String format : formats) {
            if (findFormat(format, true)) {
                sb.append(format);
            } else {
                sb.append(format);
                sb.append("%s");
            }
        }

        if (!findFormat(last, false)) {
            sb.append(last);
        }

        return sb.toString();
    }

    /**
     * This predefined FormatProcessor instance constructs a String result using {@link
     * Formatter}. Unlike {@link Formatter}, FormatProcessor uses the value from
     * the embedded expression that follows immediately after the
     * <a href="../../util/Formatter.html#syntax">format specifier</a>.
     * StringTemplate expressions without a preceeding specifier, use "%s" by

     * Example: {@snippet :
     * int x = 123;
     * int y = 987;
     * String result = FMT."%3d\{x} + %3d\{y} = %4d\{x + y}"; // @highlight substring="FMT"
     * }
     * {@link FMT} uses the Locale.US {@link Locale}.
     */
    public static final FormatProcessor FMT = new FormatProcessor(Locale.US);

}
