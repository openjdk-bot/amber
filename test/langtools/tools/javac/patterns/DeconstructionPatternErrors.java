/**
 * @test /nodynamiccopyright/
 * @summary Verify error reports for erroneous deconstruction patterns are sensible
 * @compile/fail/ref=DeconstructionPatternErrors.out --enable-preview -source ${jdk.version} -XDrawDiagnostics -XDshould-stop.at=FLOW DeconstructionPatternErrors.java
 */

import java.util.ArrayList;
import java.util.List;

public class DeconstructionPatternErrors {

    public static void main(String... args) throws Throwable {
        Object p;
        p = new P(42);
        if (p instanceof P(_));
        if (p instanceof P3(ArrayList<Integer> l));
        if (p instanceof P4(ArrayList<Integer> l));
        if (p instanceof P5(int i));
        if (p instanceof P(String s));
        if (p instanceof P5(P(var v)));
        if (p instanceof P2(var v1)); //too few nested patterns
        if (p instanceof P2(Runnable v1)); //too few nested patterns
        if (p instanceof P(var v1, var v2)); //too many nested patterns
        if (p instanceof P(int v1, int v2)); //too many nested patterns
        if (p instanceof P(int v1, Unresolvable v2)); //too many nested patterns
        if (p instanceof GenRecord<String>(var v)); //incorrect generic type
        if (p instanceof P4(GenRecord<String>(var v))); //incorrect generic type
        if (p instanceof GenRecord<String>(Integer v)); //inconsistency in types
    }

    public record P(int i) {
    }

    public record P2(Runnable r1, Runnable r2) {}
    public record P3(List<String> l) {}
    public record P4(Object o) {}
    public record P5(String s) {}
    public record GenRecord<T>(T s) {}

}
