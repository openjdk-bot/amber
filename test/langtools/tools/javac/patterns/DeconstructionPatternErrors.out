DeconstructionPatternErrors.java:15:28: compiler.err.underscore.as.identifier
DeconstructionPatternErrors.java:15:29: compiler.err.expected: token.identifier
DeconstructionPatternErrors.java:16:29: compiler.err.prob.found.req: (compiler.misc.inconvertible.types: java.util.List<java.lang.String>, java.util.ArrayList<java.lang.Integer>)
DeconstructionPatternErrors.java:17:29: compiler.err.instanceof.reifiable.not.safe: java.lang.Object, java.util.ArrayList<java.lang.Integer>
DeconstructionPatternErrors.java:18:29: compiler.err.prob.found.req: (compiler.misc.inconvertible.types: java.lang.String, int)
DeconstructionPatternErrors.java:19:28: compiler.err.prob.found.req: (compiler.misc.inconvertible.types: int, java.lang.String)
DeconstructionPatternErrors.java:20:29: compiler.err.prob.found.req: (compiler.misc.inconvertible.types: java.lang.String, DeconstructionPatternErrors.P)
DeconstructionPatternErrors.java:21:26: compiler.err.incorrect.number.of.nested.patterns: java.lang.Runnable,java.lang.Runnable, java.lang.Runnable
DeconstructionPatternErrors.java:22:26: compiler.err.incorrect.number.of.nested.patterns: java.lang.Runnable,java.lang.Runnable, java.lang.Runnable
DeconstructionPatternErrors.java:23:26: compiler.err.incorrect.number.of.nested.patterns: int, int,compiler.misc.type.none
DeconstructionPatternErrors.java:24:26: compiler.err.incorrect.number.of.nested.patterns: int, int,int
DeconstructionPatternErrors.java:25:36: compiler.err.cant.resolve.location: kindname.class, Unresolvable, , , (compiler.misc.location: kindname.class, DeconstructionPatternErrors, null)
DeconstructionPatternErrors.java:25:26: compiler.err.incorrect.number.of.nested.patterns: int, int,Unresolvable
DeconstructionPatternErrors.java:26:13: compiler.err.instanceof.reifiable.not.safe: java.lang.Object, DeconstructionPatternErrors.GenRecord<java.lang.String>
DeconstructionPatternErrors.java:27:29: compiler.err.instanceof.reifiable.not.safe: java.lang.Object, DeconstructionPatternErrors.GenRecord<java.lang.String>
DeconstructionPatternErrors.java:28:44: compiler.err.prob.found.req: (compiler.misc.inconvertible.types: java.lang.String, java.lang.Integer)
DeconstructionPatternErrors.java:28:13: compiler.err.instanceof.reifiable.not.safe: java.lang.Object, DeconstructionPatternErrors.GenRecord<java.lang.String>
- compiler.note.preview.filename: DeconstructionPatternErrors.java
- compiler.note.preview.recompile
17 errors
